README
-----------------
TtHFitter package
-----------------
Authors:
  Michele Pinamonti <michele.pinamonti@gmail.com>
  Loic Valery <loic.valery@cern.ch>
Latest Update:   01 Jul 2016
Tag:             00-01-01


---------
Setup
---------
To setup just use the script:
    source setup.sh
(should work on any machine with access to afs - provided that nothing else is set-up previously)

To compile:
    make
(this will take as main code the file util/myFit.C)
or specify a custom main file as argument:
    make  util/myNewCustomFit.C
   
      
---------
How to   
---------
To run the code, after compiling (see "Setup"), use the command:
    ./myFit.exe  <action(s)>  [<config file>]  [<update>]  [<options>]

The configuration file (<config file>) is a text file containing all the information on the definition of samples and fit regions, including all the fit and draw options.
By default the file  config/myFit.config  is loaded.
See the section "Config File" for more details.
As examples give a look at  config/myFit.config  or  config/ttH2015.config .
Most of the times the only file the user has to modify to get his fit is the configuration file. 

The only mandatory argument, <action(s)>, tells to the TtHFitter which operation(s) to perform. 
The possible operations are defined in the main file (e.g. util/myFit.C). 
For instance, if you use the default file util/myFit.C, the available options are:
    h : read input histograms (valid only if the proper option is spefified in the config file)
    n : read input ntuples (valid only if the proper option is spefified in the config file)
    w : create the RooStats xmls and workspace
    f : fit the workspace
    l : calculate exclusion limit
    s : calculate significance
    d : draw pre-fit plots
    p : draw post-fit plots
    a : draw separation plots
    r : draw ranking plot (see later)
    b : re-run smoothing (in the future also rebinning)
    m : multi-fit (see later)

New optional argument: <options>.
It's a string (so make sure to use " or ' to enclose the string if you use more than one option) defining a list of options, in the form:
    "<option1>=<value1>,<value2>,...:<option2>=..."
See the section "Command line options" below.
    

---------
Config File   
---------
Here's a list of the inputs and options which can be specifed in the config file:

 - The structure of the file shoudl be the following:

     <ObjectType>: <ObjectName>
       <ObjectProperty>: <Value>
       <ObjectProperty>: <Value>
       ...
     
     <ObjectType>: <ObjectName>
       <ObjectProperty>: <Value>
       <ObjectProperty>: <Value>
       ...

     ...

    (NB: note the blank line between the objects!!)
    
 - The file should contain:
     * exactly one object of type "Fit"
     * at least one object of type "Sample"
     * at least one object of type "Region"
     * any number of objects of type "Systematic" (even 0 should be ok)
   Each object should have unique <ObjectName>.
      
 - Then, for each object type, here's a PARTIAL list of properties to be speficied:

    Job:
      Label            -> it's the label which will be showed on the plots
      POI              -> the name of the parameter of interest; this should correspond to a NormFactor defined below
      ReadFrom         -> can be HIST or NTUP; default is HIST
      HistoPath        -> valid only for option HIST above is selected; it's the path where the input root files containing the histograms are stored
      NtuplePath(s)    -> valid only for option NTUP; it's the path(s) where the input root files containing the ntuples are stored
      MCweight         -> only for option NTUP; string defining the weight (for MC samples only)
      Selection        -> only for option NTUP; string defining the selection
      NtupleName       -> default name of the tree
      Lumi             -> value to scale all the "NormalizedByTheory" samples
      LumiScale        -> additional value to scale 'after' histogram creation (for fast scaling) IMPORTANT: use it only if you know what you are doing!!
      SystPruningShape -> Lower threshold to remove a shape systematic from the fit/limit (suppression is done per sample and per region) (Ex: 0.02 for 2%)
      SystPruningNorm  -> Lower threshold to remove a normalisation systematic from the fit/limit (suppression is done per sample and per region) (Ex: 0.02 for 2%)
      SystLarge        -> all systematics above this threshold will be flagged in the pruning plot) (e.g. 0.4 will flag systematics that are larger than 40%)
      IntCodeOverall   -> interpolation code used for the normalization component of systematics (should match the one used in RooStats)
      IntCodeShape     -> interpolation code used for the shape component of systematics (should match the one used in RooStats)
      MCstatThreshold  -> if set it will add the MC stat uncertainty to the fit (and to the plots); a NP will be added for each bin with an MC stat uncertainty > this threshold (relative)
      DebugLevel       -> 0 or 1
      PlotOptions      -> a set os options for plotting:
        YIELDS : if set, the legend will be one-column and will include the yileds; otherwise two-columns and no yields
        NORMSIG : add normlised signal to plots
        NOSIG: don't show signal in stack
        OVERSIG: overlay signal (not normalised)
      SystControlPlots -> if set to true, plots will be dumped showing the shape effect of a given systematic (before and after smoothing/symmetrisation)
      SystDataPlots    -> if set to true, plots will be dumped showing the shape effect of a given systematic (before and after smoothing/symmetrisation) on top of the nominal sum of samples.
      		          Data are then plotted in the ratio. If the option is set to "fillUpFrame" data will also be plotted in the upper frame.
      CorrelationThreshold -> Threshold used to draw the correaltion matrix (only systematics with at least one correlation larger than than draw) (0.05->5%)
      SignalRegionsPlot-> list of regions to put in SignalRegionsPlot; use "EMPTY" to put an empty entry, "ENDL" to specify end of line
      HistoChecks      -> NOCRASH: means that if an error is found in the input histograms, the code continues (with only warnings) -- default leads to a crash in case of problem
      LumiLabel        -> label for luminosity to be put on plots
      CmeLabel         -> label for center-of-mass energy to be put on plots
      SplitHistoFiles  -> set this to true to have histogram files split by region (useful with many regions and/or run in parallel)
      BlindingThreshold-> bins with S/B > this number will be blinded
      RankingMaxNP     -> max number of NP to show in ranking plot
      RankingPlot      -> NP categories in gammas or systs, if set to Systs(Gammas) then plot only systs(Gammas) in ranking, default produce plot for systs+gammas, can also set to all to have the 3 plots.
      ImageFormat      -> png, pdf or eps
      StatOnly         -> the code ignores the MC stat uncertainties from all computations (limits, significances, fit, ...)
      SystErrorBars    -> add stat error bars to syst variations in syst plots
      SummaryPlotRegions -> list of regions to be shown in summary plot (useful to specify a custom order)
      FixNPforStatOnly -> if set to TRUE, when running stat-only (with either of the two options) also the norm facotrs other than the POI are kept fixed
      InputFolder      -> specify it to read fit input histograms from a different direcotry than <jobName>/Histograms/
      InputName        -> specify it to read fit input histograms from files with different name than <jobName>_blabla.root
      WorkspaceFileName -> if specified, an external ws can be used as input for fitting (not 100% supported)
      KeepPruning      -> if set to TRUE, the first time the ws is created (option w) a Pruning.root file is created under <jobName>/ and used for future operations to skip pruned systematics (makes operations much faster in case many syst are pruned)
      AtlasLabel       -> to specify Internal, Preliminary, etc...
      CleanTables      -> if set to TRUE, a cleaned version of the tex tables is created (basically removing the "#") - to be expanded
      SystCategoryTables -> if set to TRUE, additional syst tables with systematics grouped by category are created

    Fit:
      FitType          -> can be SPLUSB (default) or BONLY to fit under the s+b or the b-only hypothesis
      FitRegion        -> can be CRSR (default) or CRONLY to fit considering both signal and control regions in the fit, or only control regions
                          You can also specify a coma-separated list of regions to use in the fit
      FitBlind         -> specify is real data or Asimov data should be used in the fit (TRUE or FALSE). By default, fit are NOT blind.
      POIAsimov        -> value of the parameter of interest in the AsimovDataset used in the fit
      NPValues         -> values of the nuisance parameters used to build the fit. Coma-separated list of NP:value (e.g. alpha_ttbarbb_XS:1,alpha_ttbarbcc_XS:1.5)
      doLHscan         -> coma separated list of names of the POI or NP from which you want to produce the likelihood scan, if first element of the list is "all" then all systematics are profiled
      UseMinos         -> coma separated list of names of the POI and/or NP for which you want to calculate the MINOS errors, if first element of the list is "all" then the MINOS errors is calculated for all systematics and POIs
      SetRandomInitialNPval -> useful to set this to >0 (e.g. 0.1) to help convergence of Asimov fits
      NumCPU           -> specify the number of CPU to use for the minimization (default = 1)
      StatOnlyFit      -> if specified, the fit will keep fixed all the NP to the latest fit result, and the fit results will be saved with the _statOnly suffix (also possible to use it from command line)

    Limit:
      LimitType        -> can be ASYMPTOTIC or TOYS (the latter is not yet supported)
      LimitBlind       -> can be TRUE or FALSE (TRUE means that ALL regions are blinded)
      POIAsimov        -> value of the POI to inject in the Asimov dataset in LimitBlind is set to TRUE
      
    Options:           (additional options, accepting only float as arguments - useful for adding your functionalities & flags in a quick way, since they need minimal changes in the code)
      ...
      
    Region:
      VariableTitle    -> it's the label which will be displayed on the x-axis in the plots
      Label            -> it's the label which will be showed on the plots and specifies which region is shown
      TexLabel         -> label for tex files
      ShortLabel       -> same as above, but a shorter version for plots with smaller available place
      LumiLabel        -> label for luminosity to be put on plots
      CmeLabel         -> label for center-of-mass energy to be put on plots
      LogScale         -> set it to true to have log-scale when plotting this region
      HistoFile        -> only for option HIST, the file name to be used
      HistoName        -> only for option HIST, the histogram name to be used
      HistoPathSuff(s) -> only for option HIST, the path suffix (or suffixes, comma-separated) where to find the histogram files for this region
      Variable         -> only for option NTUP, the variable (or expression) inside the ntuple to plot
      		          can define a variable as X|Y to do the correlation plot between X and Y
      Selection        -> only for option NTUP, the selection done on the ntuple for this region
      NtupleName       -> only for option NTUP, the name of the tree for this region
      NtuplePathSuff(s)-> only for option NTUP, the path sufix (or suffixes, comma-separated) where to find the ntuple files for this region
      MCweight         -> only for option NTUP, the additional weight sed in this region (for MC samples only)
      Rebin            -> if specified, the histograms will be rebinned merging N bins together, where N is the argument (int)
      Binning          -> if specified, the histograms will be rebinned according to the new binning specifed, in the form like (0,10,20,50,100)
      		       	  if option AutoBin is set, use algorithms/functions ro define the binning. Example - Binning: "AutoBin","TransfoD",5.,6. (TransfoF also available, 5. and 6. are parameters of the transformation)
			  if used in background region and zSig!=0 (first parameter, =0 gives flat background) then need a coma separated list of backgrounds to use instead of signal to compute the binning.
      BinWidth         -> if specified, two things are done: this number is used to decorate the y axis label and the bin content is scaled for bins with a bin width different from this number
      Type             -> can be SIGNAL, CONTROL or VALIDATION; used depending on Fit->FitType; if VALIDATION is set, the region is never fitted; default is SIGNAL
      DataType         -> ASIMOV or DATA. Is asimov is set, the limits and significances are computed without taking into acount the data in these region, but a projection of the fit performed in the regions with DATA

    Sample:
      Type             -> can be SIGNAL, BACKGROUND, DATA or GHOST; default is BACKGROUND; GHOST means: no syst, not drawn, not propagated to workspace
      Title            -> title shown on the legends
      TexTitle         -> title shown on tex tables
      Group            -> if specified, sample will be grouped with other samples with same group and this label will be used in plots
      HistoFile        -> valid only for option HIST; which root file to read (excluding the suffix ".root"); this will be combined with Fit->HistoPath to build the full path
      HistoName        -> valid only for option HIST; name of histogram to read
      NtuplePath       -> valid only for option HIST; it's the path where the input root files containing the histograms are stored
      NtupleFile(s)    -> valid only for option NTUP; it's the file name(s) where the input ntuples are stored
      NtupleName       -> valid only for option NTUP; name of tree to read
      NtuplePath(s)    -> valid only for option NTUP; it's the path(s) where the input root files containing the ntuples are stored
      FillColor        -> histogram fill color (not valid for data)
      LineColor        -> histogram line color
      NormFactor       -> NormalisationFactor (free parameter in the fit); in the format <name>,nominal,min,max
      NormalizedByTheory-> set it to false for data-driven backgrounds (MCweight, Lumi and LumiScale from Job and Region will be ignored)
      MCweight         -> only for option NTUP, the additional weight sed in this sample (for all types of samples!! Not only MC)
      Selection        -> valid only for option NTUP; additional selection for this region
      Regions          -> set this to have the sample only in some regions
      Exclude          -> set this to exclude the sample in some regions
      LumiScale(s)     -> set this to scale the sample by a number; if more numbers are set, use a different one for each file / name / path...
      IgnoreSelection  -> if set, selection from Job and Region will be ignored
      UseMCstat        -> if set to FALSE, makes the fitter ignore the stat uncertainty for this sample
      MultiplyBy       -> if specified, each sample hist is multiplied bin-by-bin by another sample hist, in each of the regions
      DivideBy         -> if specified, each sample hist is divided bin-by-bin by another sample hist, in each of the regions
      
    NormFactor:
      Samples          -> comma-separated list of samples on which to apply the norm factor
      Regions          -> comma-separated list of regions where to apply the norm factor
      Exclude          -> comma-separated list of samples/regions to exclude
      Title            -> title of the norm factor
      Nominal          -> nominal value
      Min              -> min value
      Max              -> max value
      Constant         -> set to TRUE to have a fixed norm factor 
      
    Systematic:
      Samples          -> comma-separated list of samples on which to apply the systematic
      Regions          -> comma-separated list of regions where to apply the systematic
      Exclude          -> comma-separated list of samples/regions to exclude
      Type             -> can be HISTO or OVERALL
      Title            -> title of the systematic (will be shown in plots)
      Category         -> major category to which the systematic belongs (instrumental, theory, ttbar, ...): used to split pulls plot for same category
      HistoPathUp      -> only for option HIST, for HISTO systematic: histogram file path for systematic up variation
      HistoPathDown    -> only for option HIST, for HISTO systematic: histogram file path for systematic down variation
      HistoPathSufUp   -> only for option HIST, for HISTO systematic: suffix of the histogram file names for systematic up variation
      HistoPathSufDown -> only for option HIST, for HISTO systematic: suffix of the histogram file names for systematic down variation
      HistoFileUp      -> only for option HIST, for HISTO systematic: histogram file name for systematic up variation
      HistoFileDown    -> only for option HIST, for HISTO systematic: histogram file name for systematic down variation
      HistoFileSufUp   -> only for option HIST, for HISTO systematic: suffix of the histogram file names for systematic up variation
      HistoFileSufDown -> only for option HIST, for HISTO systematic: suffix of the histogram file names for systematic down variation
      HistoNameUp      -> only for option HIST, for HISTO systematic: histogram name for systematic up variation
      HistoNameDown    -> only for option HIST, for HISTO systematic: histogram name for systematic down variation
      HistoNameSufUp   -> only for option HIST, for HISTO systematic: suffix of the histogram names for systematic up variation
      HistoNameSufDown -> only for option HIST, for HISTO systematic: suffix of the histogram names for systematic down variation
      NtuplePathsUp    -> only for option NTUP, for HISTO systematic: ntuple file path for systematic up variation
      NtuplePathsDown  -> only for option NTUP, for HISTO systematic: ntuple file path for systematic down variation
      NtuplePathSufUp  -> only for option NTUP, for HISTO systematic: suffix of the ntuple file paths for systematic up variation
      NtuplePathSufDown-> only for option NTUP, for HISTO systematic: suffix of the ntuple file paths for systematic down variation
      NtupleFilesUp    -> only for option NTUP, for HISTO systematic: ntuple file name for systematic up variation
      NtupleFilesDown  -> only for option NTUP, for HISTO systematic: ntuple file name for systematic down variation
      NtupleFileSufUp  -> only for option NTUP, for HISTO systematic: suffix of the ntuple file names for systematic up variation
      NtupleFileSufDown-> only for option NTUP, for HISTO systematic: suffix of the ntuple file names for systematic down variation
      NtupleNamesUp    -> only for option NTUP, for HISTO systematic: ntuple name for systematic up variation
      NtupleNamesDown  -> only for option NTUP, for HISTO systematic: ntuple name for systematic down variation
      NtupleNameSufUp  -> only for option NTUP, for HISTO systematic: suffix of the ntuple names for systematic up variation
      NtupleNameSufDown-> only for option NTUP, for HISTO systematic: suffix of the ntuple names for systematic down variation
      WeightUp         -> only for option NTUP, for HISTO systematic: weight for systematic up variation
      WeightDown       -> only for option NTUP, for HISTO systematic: weight for systematic down variation
      WeightSufUp      -> only for option NTUP, for HISTO systematic: additional weight for systematic up variation
      WeightSufDown    -> only for option NTUP, for HISTO systematic: additional weight for systematic down variation
      IgnoreWeight     -> only for option NTUP: if set, the corresponding weight (present in Job, Sample or Region) will be ignored for this systematic
      Symmetrisation   -> can be ONESIDED or TWOSIDED (...); for no symmetrisation, skip the line
      Smoothing        -> smoothing code to apply; use 40 for default smoothing; for no smoothing, skip the line
      OverallUp        -> for OVERALL systematic: the relative "up" shift (0.1 means +10%)
      OverallDown      -> for OVERALL systematic: the relative "down" shift (-0.1 means -10%)
      ReferenceSample  -> if this is specified, the syst variation is evaluated w.r.t. this reference sample (often a GHOST sample) instead of the nominal, and then the relative difference is propagated to nominal; NOTE: also the overall relative difference is propagated


---------
Command line options
---------

Currently the supported options are:
      Regions:     to limit the regions to use to the list specified
      Samples:     to limit the samples to use to the list specified
      Systematics: to limit the systematics to use to the list specified
      Signal:      in case more than one SIGNAL sample is specified in your config file, you can specify which one you want to run on (for plots, workspace creation and fits/limits/significance)
      Exclude:     to exclude certain Regions / Samples / Systematics
      Suffix:      used for: plots, workspace, fit resutls, etc
      Update:      if TRUE, the output .root file is updated, otherwise is overwrote
      StatOnlyFit: if TRUE, the same as Fit, StatOnlyFit
Note: the wild-card * is supported, but only as last character.
Example:
      ./myFit.exe  n  config/ttH2015.config 'Regions=HThad_ge6jge4b;Exclude=BTag_*'


---------
Ranking Plot
---------

 - The ranking plot can be created in one go, with just the command line argument "r" (after having run the nominal fit fit "f").
 - Since this can take too much time (and memory), for complicated fits it's better to run it in several steps:
   by specifying the command-line option "Ranking=<name/index>"
   one can produce the txt input for the ranking only for a specific line of the ranking, i.e. for a single NP (speficied either through its name or index).
   Once all the needed txt files are created (e.g. in parallel throgh batch jobs) with the option "Ranking=plot" they are merged to create the final plot.
 - Examples:
     # this runs the ranking in one go
     ./myFit.exe  r  <config>
     #these commands will first create the inputs for the ranking one by one and then merge them in the plot
     ./myFit.exe  r  <config> Ranking=Lumi
     ./myFit.exe  r  <config> Ranking=JES1
     ./myFit.exe  r  <config> Ranking=ttXsec
     ./myFit.exe  r  <config> Ranking=plot


---------
Multi-Fit
---------

The Multi-Fit functionality can be sued to compare fit results or even to combine fit inputs from different configuration files / Jobs.
 - To use it you need a dedicated config file, with a similar starucure as the usual ones. Example:
    
  ---
  file: config/myTopWS_multifit.config
  ---
    MultiFit: "myTopWS_multifit"
      Label: "My Lable"
      Combine: FALSE
      Compare: TRUE
      CmeLabel: "13 TeV"

ChangeLog
---------------------------------------------------------------------------------------
The up-to date changelog can be found here:
https://docs.google.com/document/d/1Gcp06Rfc0e09GRQrwRMgQLqKisTnKoeei7sZlG4a-Qo/preview
The changelog version below was last synchronized on 2019-07-25.
---------------------------------------------------------------------------------------

2019-07-25 Alexander Held <alexander.held@cern.ch>
    * Updating ChangeLog to current state of google docs, adding link to google docs version.

2019-07-09 Tomas Dado <tomas.dado@cern.ch>
    * Fixing bug in Configreader definition of fLHScanStepsY. MR prepared by Nicolas Scharmberg, thanks!
    * Fixing problem with postfit plots when morphing is used with SaturatedModel
    * Adding support for the setup that allows to run TRExFitter within ATLAS framework (AnalysisBase/AnalysisTop). MR prepared by  Nicolas Scharmberg, thanks!

2019-07-05 Tomas Dado <tomas.dado@cern.ch>
    * Added macro to plot delta-POI

2019-07-05 Michele Pinamonti <michele.pinamonti@gmail.com>
    * Fix to post-fit syst band when using morphing (and with more systematics with same NuisanceParameter)

2019-07-04 Tomas Dado <tomas.dado@cern.ch>
    * Fixing crash during plotting when only some regions were used with SignalRegionsPlot
    * Fixing problems with wildcards in ntuple paths reported in TTHFITTER-238
    * Updated CommonStatTools and CommonSystSmoothingTools to the latest version

2019-06-27 Tomas Dado <tomas.dado@cern.ch>
    * Fixing histos names for pseudodata. MR prepared by  Nicolas Scharmberg, thanks!
    * Adding option to run 2D likelihood scan. MR prepared by  Nicolas Scharmberg, thanks!
    * Properly propagating NF to correct regions in Pre-fit.

2019-06-24 Tomas Dado <tomas.dado@cern.ch>
    * Updating ROOT release to 6.14.04

===================================================
TAG TtHFitter-00-04-06
===================================================

2019-06-20 Michele Pinamonti <michele.pinamonti@gmail.com>
    * Fix to syst band in pre and post fit (for specific situations)

2019-06-20 Tomas Dado <tomas.dado@cern.ch>
    * Fixing decorrelation between samples. MR prepared by FIlip Nechansky, thanks!

2019-06-18 Tomas Dado <tomas.dado@cern.ch>
    * Switching to binned likelihood fit when running toys. MR prepared by Kevin Moor, thanks!

2019-06-17 Tomas Dado <tomas.dado@cern.ch>
    * Fixing a bug affecting normalisation factors in ranking. MR prepared by Ian Connelly, thanks!
    * Including syst uncertainties not present in the correlation matrix to syst band. MR prepared by Giancarlo Panizzo, thanks!

2019-06-06 Tomas Dado <tomas.dado@cern.ch>
    * Added Bootstrap command line option
    * MR prepared by Filip Nechansky, thanks!

2019-05-31 Tomas Dado <tomas.dado@cern.ch>
    * Fixing Divide/Multiply/Subtract working again for samples
    * Postfit plotting speed up
    * MR prepared by Giancarlo Panizzo, thanks!

2019-05-31 Michele Pinamonti <michele.pinamonti@gmail.com>
    * Added MaxNtupleEvents option (for NTUP)
    * Added SystFromSample Sample option
    * Fixed post-fit plots so that systematics not in fit results are still used and added to band
    * Other small fixes

2019-05-28 Tomas Dado <tomas.dado@cern.ch>
    * Added option to run Boostrapping for some systematics during the “n” step. MR prepared by Filip Nechansky, thanks!

2019-05-27 Alexander Held <alexander.held@cern.ch>
    * Added some “hidden” options to the jobSchema

2019-05-27 Tomas Dado <tomas.dado@cern.ch>
    * Fixing issues with Float to Pseudohex that ignored leading zeros. MR prepared by Kevin Moor, thanks!

2019-05-27 Tomas Dado <tomas.dado@cern.ch>
    * When using MergeUnderflowOverflow setting the errors of underflow/overflow to zero as it was causing issues with smoothing. MR prepared by Nicolas Scharmberg, thanks!

2019-05-15 Tomas Dado <tomas.dado@cern.ch>
    * Multifit: check if the config files used are not identical (would cause problems)

===================================================
TAG TtHFitter-00-04-05
===================================================

2019-05-09 Tomas Dado <tomas.dado@cern.ch>
    * Fixing SaturatedModel interfering with chi^2 calculation

2019-04-24 Michele Pinamonti <michele.pinamonti@gmail.com>
    * Added new config option to a dummy systematics for some samples (syst variation = nominal) - useful when used in combination with KeepNormForSamples

2019-04-24 Giancarlo Panizzo <giancarlo.panizzo@cern.ch>
    * Implementing “IgnoreWeight” config option for systematics (didn't do anything before) (Issue #8)

2019-04-19 Giancarlo Panizzo <giancarlo.panizzo@cern.ch>
    * AddSamples / SubtractSamples/DivideBy/MultiplyBy now inherit systematics by NuisanceParameter

2019-04-19 Alexander Held <alexander.held@cern.ch>
    * Added nicer colour scheme for correlation matrix

2019-04-16 Alexander Held <alexander.held@cern.ch>
    * More fixes for missing files

2019-04-16 Tomas Dado <tomas.dado@cern.ch>
    * Fixing usage of wildcards in Sample option (Issue #7)
    * Now properly printing errors/warnings when input files for ntuple reading are not found
    * Improved LHscan fitting by using Migrad directly - results in smoother LH curves
    * Fixing scaling of Morphing samples when doing S/B plots

2019-04-15 Tomas Dado <tomas.dado@cern.ch>
    * Fixing consistent usage of ttbarresonance option when smoothing nominal samples

2019-04-12 Tomas Dado <tomas.dado@cern.ch>
    * Speed improvements of the plotting code (postfit)

2019-04-12 Michele Pinamonti <michele.pinamonti@gmail.com>
    * Added saturated model for goodness of fit estimation

2019-03-28 Michele Pinamonti <michele.pinamonti@gmail.com>
    * Command-line “Signals” added/changed
    * Job option “ScaleSamplesToData” added (allows to scale a set of samples to data minus the rest when plotting (d option) )
    * Job option “ExcludeFromMorphing” added
    * Other small fixes

2019-03-28 Tomas Dado <tomas.dado@cern.ch>

Fit4t.C
#include "TFile.h"
#include "TH1F.h"
#include "TCanvas.h"

#include "Root/Common.C"
#include "Root/NuisParameter.C"
#include "Root/CorrelationMatrix.C"
#include "Root/FitResults.C"
#include "Root/Systematic.C"
#include "Root/SystematicHist.C"
#include "Root/NormFactor.C"
#include "Root/Sample.C"
#include "Root/Region.C"
#include "Root/SampleHist.C"
#include "Root/TtHFit.C"

#include "Root/TthPlot.C"


// opt:
// "t" -> read ntuples
// "w" -> create workspace
// "f" -> fit
// "d" -> draw pre-fit
// "p" -> draw post-fit

void Fit4t(string opt="t",bool update=false){
  bool useTRF = true;
  
  TtHFitter::SetDebugLevel(1);    

  // interpret opt
  bool readNtuples     = opt.find("t")!=string::npos;
  bool createWorkspace = opt.find("w")!=string::npos;
  bool doFit           = opt.find("f")!=string::npos;
  bool doLimit         = opt.find("l")!=string::npos;
  bool drawPreFit      = opt.find("d")!=string::npos;
  bool drawPostFit     = opt.find("p")!=string::npos;
  bool systSmoothing   = opt.find("s")!=string::npos;
  
  // create the fit object
  TtHFit *myFit = new TtHFit();
    // ntuple stuff
    myFit->AddNtuplePath("/gpfs/atlas/public/TopMiniSLs/TRC_23/NTUP_ELE_V1_L2/nomin/");
    myFit->AddNtuplePath("/gpfs/atlas/public/TopMiniSLs/TRC_23/NTUP_MUON_V1_L2/nomin/");
    if(!useTRF) myFit->SetMCweight("FinalWeight * weight_BTag");
    else myFit->SetMCweight("FinalWeight");
    myFit->SetSelection("(lep_trigMatch&3)>0");
    myFit->SetNtupleName("mini");
    
  // create the samples
  Sample *data = myFit->NewSample("Data",SampleType::Data);
    data->SetTitle("Data 2012");
    data->AddNtupleFile("data");
    
  Sample *ttH = myFit->NewSample("ttH",SampleType::Background);
    ttH->SetTitle("t#bar{t}+H");
    ttH->SetFillColor(kGray);
    ttH->SetLineColor(kBlack);
    ttH->AddNtupleFile("ttH");
  Sample *ttV = myFit->NewSample("ttV",SampleType::Background);
    ttV->SetTitle("t#bar{t}+W/Z");
    ttV->SetFillColor(kGray+2);
    ttV->SetLineColor(kBlack);
    ttV->AddNtupleFile("ttbar_V");

  Sample *ttbar = myFit->NewSample("ttbar",SampleType::Background);
    ttbar->SetTitle("t#bar{t}+jets");
    ttbar->SetFillColor(kWhite);
    ttbar->SetLineColor(kBlack);
    ttbar->AddNtupleFile("ttbar_hdampHeraPdf");
//     ttbar->AddNtupleFile("ttbar_sherpa");
//     ttbar->AddNtupleFile("ttbar_MadGraph");

    // norm uncertainties
    ttbar->AddSystematic("ttXsec",SystType::Overall,0.10,-0.10);
//     ttbar->AddSystematic("ttXsec",SystType::Overall,0.0001,-0.0001);
//     ttH->AddSystematic("tthXsec",SystType::Overall,0.50,-0.50);
//     ttV->AddSystematic("ttvXsec",SystType::Overall,0.50,-0.50);

    // 24% uncertainty every jet bin above 6 (kind of Berend scaling)
    Systematic *ttJetRatio = ttbar->AddSystematic("ttJetRatio",SystType::Histo);
      ttJetRatio->fWeightSufUp   = "(1+(jet_n - 4)*0.05)";
      ttJetRatio->fWeightSufDown = "(1-(jet_n - 4)*0.05)";

    Systematic *ttMCsherpa = ttbar->AddSystematic("ttMCsherpa",SystType::Histo);
      ttMCsherpa->fNtupleFilesUp.push_back("ttbar_sherpa");
    Systematic *ttMCmg = ttbar->AddSystematic("ttMCmg",SystType::Histo);
      ttMCmg->fNtupleFilesUp.push_back("ttbar_MadGraph");
      
  Sample *sig = myFit->NewSample("Signal",SampleType::Signal);
    sig->SetTitle("SM t#bar{t}t#bar{t}");
    sig->SetFillColor(kRed);
    sig->SetLineColor(kRed);
    sig->AddNormFactor("SigXsecOverSM",1,0,100);
    sig->AddNtupleFile("tttt_SM");
    

  // control region
  Region *CR_ge4j2b = myFit->NewRegion("CR_ge4j2b");
    if(!useTRF) CR_ge4j2b->AddSelection("jet_n>=4 && nJet_tagged==2");
    else{
      CR_ge4j2b->AddSelection("jet_n>=4 && (channelNumber!=runNumber || nJet_tagged==2)");
      CR_ge4j2b->AddMCweight("weight_TRF_2b");
    }
    CR_ge4j2b->SetVariable("jet_n",7,4,11);
    CR_ge4j2b->SetVariableTitle("Number of jets");
    CR_ge4j2b->SetLabel("e/#mu + #geq4 j, 2 b", "#geq4 j, 2 b");
    
  // signal regions
  Region *SR_9jge4b = myFit->NewRegion("SR_9jge4b");
    if(!useTRF) SR_9jge4b->AddSelection("jet_n==9 && nJet_tagged>=4");
    else{
      SR_9jge4b->AddSelection("jet_n==9 && (channelNumber!=runNumber || nJet_tagged>=4)");
      SR_9jge4b->AddMCweight("weight_TRF_ge4b");
    }
    SR_9jge4b->SetVariable("(HTj+lep_pt+met_et)/1e3",7,400,1800);
    SR_9jge4b->SetVariableTitle("H_{T} [GeV]");
    SR_9jge4b->SetLabel("e/#mu + 9 j, #geq4 b", "9 j, #geq3 b");

  Region *SR_ge10j3b = myFit->NewRegion("SR_ge10j3b");
    if(!useTRF) SR_ge10j3b->AddSelection("jet_n>=10 && nJet_tagged==3");
    else{
      SR_ge10j3b->AddSelection("jet_n>=10 && (channelNumber!=runNumber || nJet_tagged==3)");
      SR_ge10j3b->AddMCweight("weight_TRF_3b");
    }
    SR_ge10j3b->SetVariable("(HTj+lep_pt+met_et)/1e3",7,400,1800);
    SR_ge10j3b->SetVariableTitle("H_{T} [GeV]");
    SR_ge10j3b->SetLabel("e/#mu + #geq10 j, 3 b", "#geq10 j, 3 b");
    
  Region *SR_ge10jge4b = myFit->NewRegion("SR_ge10jge4b");
    if(!useTRF) SR_ge10jge4b->AddSelection("jet_n>=10 && nJet_tagged>=4");
    else{
      SR_ge10jge4b->AddSelection("jet_n>=10 && (channelNumber!=runNumber || nJet_tagged>=4)");
      SR_ge10jge4b->AddMCweight("weight_TRF_ge4b");
    }
    SR_ge10jge4b->SetVariable("(HTj+lep_pt+met_et)/1e3",7,400,1800);
    SR_ge10jge4b->SetVariableTitle("H_{T} [GeV]");
    SR_ge10jge4b->SetLabel("e/#mu + #geq10 j, #geq4 b", "#geq10 j, #geq4 b");
    
    
  //
  // do actual things
  //
  if(readNtuples){
    myFit->ReadNtuples();
    myFit->Print();
    myFit->WriteHistos("",!update);
  }
  else{
    myFit->ReadHistos("");
  }
  
  if(systSmoothing){
    myFit->SmoothSystematics("all");
    myFit->WriteHistos("",!update);
  }
  
  if(drawPreFit){
    myFit->DrawAndSaveAll();
//     myFit->DrawAndSaveAll("blind");
//     myFit->DrawSystPlots();
    myFit->DrawSummary("log");
    Region *regions[4];
      regions[0] = CR_ge4j2b;
      regions[1] = SR_9jge4b;
      regions[2] = SR_ge10j3b;
      regions[3] = SR_ge10jge4b;
    myFit->DrawSignalRegionsPlot(2,2,regions);
  }

  if(createWorkspace){
    myFit->SetPOI("SigXsecOverSM");
//     myFit->SetLumiErr(0.037);
    myFit->SetLumiErr(0.);
//     myFit->SetStatErrorConfig(true,0.05);
    myFit->SetStatErrorConfig(false,0.05);
    myFit->ToRooStat(true,true);
//     myFit->ToRooStat(true,false);
  }

  // use the external tool FitCrossCheckForLimits fir fitting
  if(doFit){
    myFit->Fit(); // with FitCrossCheckForLimits
    myFit->PlotFittedNP();
  }
  
  if(doLimit){
    myFit->GetLimit();
  }
  
  if(drawPostFit){
//     myFit->ReadFitResults("xcheckResults/TextFileFitResult/GlobalFit_fitres_unconditionnal_mu0.txt");
//     SR_ge10jge4b->DrawPostFit(myFit->fFitResults)->SaveAs("Test.png");
    myFit->DrawAndSaveAll("post");
  }
  
}
 
